File: internal/notify/notify.go
package notify

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
	"time"
)

type WebhookType string

const (
	WebhookDiscord WebhookType = "discord"
	WebhookSlack   WebhookType = "slack"
	WebhookGeneric WebhookType = "generic"
)

type CompleteOptions struct {
	SessionName string
	WebhookURL  string
	ProjectDir  string
	Iterations  int
	Duration    time.Duration
	Timeout     time.Duration
}

type FailedOptions struct {
	SessionName    string
	WebhookURL     string
	FailureReason  string
	ProjectDir     string
	Iterations     int
	MaxIterations  int
	RemainingTasks int
	Duration       time.Duration
	Timeout        time.Duration
}

func DetectWebhookType(url string) WebhookType {
	lower := strings.ToLower(url)
	if strings.Contains(lower, "discord.com/api/webhooks") || strings.Contains(lower, "discordapp.com/api/webhooks") {
		return WebhookDiscord
	}
	if strings.Contains(lower, "hooks.slack.com") {
		return WebhookSlack
	}
	return WebhookGeneric
}

func NotifyComplete(ctx context.Context, opts CompleteOptions) error {
	if strings.TrimSpace(opts.SessionName) == "" {
		return errors.New("session name is required")
	}
	if strings.TrimSpace(opts.WebhookURL) == "" {
		return errors.New("webhook URL is required")
	}
	payload, err := buildCompletePayload(opts, time.Now())
	if err != nil {
		return err
	}
	return SendWebhook(ctx, opts.WebhookURL, payload, opts.Timeout)
}

func NotifyFailed(ctx context.Context, opts FailedOptions) error {
	if strings.TrimSpace(opts.SessionName) == "" {
		return errors.New("session name is required")
	}
	if strings.TrimSpace(opts.WebhookURL) == "" {
		return errors.New("webhook URL is required")
	}
	payload, err := buildFailedPayload(opts, time.Now())
	if err != nil {
		return err
	}
	return SendWebhook(ctx, opts.WebhookURL, payload, opts.Timeout)
}

func SendWebhook(ctx context.Context, url string, payload []byte, timeout time.Duration) error {
	if strings.TrimSpace(url) == "" {
		return errors.New("webhook URL is required")
	}
	if len(payload) == 0 {
		return errors.New("payload is required")
	}
	if timeout <= 0 {
		timeout = 30 * time.Second
	}
	if ctx == nil {
		ctx = context.Background()
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(payload))
	if err != nil {
		return fmt.Errorf("create webhook request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: timeout}
	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("send webhook: %w", err)
	}
	defer resp.Body.Close()
	_, _ = io.Copy(io.Discard, resp.Body)

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("webhook returned HTTP %d", resp.StatusCode)
	}
	return nil
}

func buildCompletePayload(opts CompleteOptions, now time.Time) ([]byte, error) {
	project := defaultString(opts.ProjectDir, "unknown")
	iterations := numberString(opts.Iterations)
	duration := formatDuration(opts.Duration)
	timestamp := now.Format(time.RFC3339)

	payloadType := DetectWebhookType(opts.WebhookURL)
	switch payloadType {
	case WebhookDiscord:
		payload := map[string]interface{}{
			"embeds": []map[string]interface{}{
				{
					"title":       "\u2705 Gralph Complete",
					"description": fmt.Sprintf("Session **%s** has finished all tasks successfully.", opts.SessionName),
					"color":       5763719,
					"fields": []map[string]interface{}{
						{
							"name":   "Project",
							"value":  fmt.Sprintf("`%s`", project),
							"inline": false,
						},
						{
							"name":   "Iterations",
							"value":  iterations,
							"inline": true,
						},
						{
							"name":   "Duration",
							"value":  duration,
							"inline": true,
						},
					},
					"footer": map[string]interface{}{
						"text": "Gralph CLI",
					},
					"timestamp": timestamp,
				},
			},
		}
		return json.Marshal(payload)
	case WebhookSlack:
		payload := map[string]interface{}{
			"attachments": []map[string]interface{}{
				{
					"color": "#57F287",
					"blocks": []map[string]interface{}{
						{
							"type": "header",
							"text": map[string]interface{}{
								"type":  "plain_text",
								"text":  "\u2705 Gralph Complete",
								"emoji": true,
							},
						},
						{
							"type": "section",
							"text": map[string]interface{}{
								"type": "mrkdwn",
								"text": fmt.Sprintf("Session *%s* has finished all tasks successfully.", opts.SessionName),
							},
						},
						{
							"type": "section",
							"fields": []map[string]interface{}{
								{
									"type": "mrkdwn",
									"text": fmt.Sprintf("*Project:*\n`%s`", project),
								},
								{
									"type": "mrkdwn",
									"text": fmt.Sprintf("*Iterations:*\n%s", iterations),
								},
								{
									"type": "mrkdwn",
									"text": fmt.Sprintf("*Duration:*\n%s", duration),
								},
							},
						},
						{
							"type": "context",
							"elements": []map[string]interface{}{
								{
									"type": "mrkdwn",
									"text": fmt.Sprintf("Gralph CLI \u2022 %s", timestamp),
								},
							},
						},
					},
				},
			},
		}
		return json.Marshal(payload)
	default:
		payload := map[string]interface{}{
			"event":      "complete",
			"status":     "success",
			"session":    opts.SessionName,
			"project":    project,
			"iterations": iterations,
			"duration":   duration,
			"timestamp":  timestamp,
			"message":    fmt.Sprintf("Gralph loop '%s' completed successfully after %s iterations (%s)", opts.SessionName, iterations, duration),
		}
		return json.Marshal(payload)
	}
}

func buildFailedPayload(opts FailedOptions, now time.Time) ([]byte, error) {
	project := defaultString(opts.ProjectDir, "unknown")
	reason := defaultString(opts.FailureReason, "unknown")
	iterations := numberString(opts.Iterations)
	maxIterations := numberString(opts.MaxIterations)
	remaining := numberString(opts.RemainingTasks)
	duration := formatDuration(opts.Duration)
	timestamp := now.Format(time.RFC3339)

	slackDescription := failedDescription(reason, opts.SessionName, true)
	discordDescription := failedDescription(reason, opts.SessionName, false)
	message := failedMessage(reason, opts.SessionName, iterations, maxIterations, remaining)

	payloadType := DetectWebhookType(opts.WebhookURL)
	switch payloadType {
	case WebhookDiscord:
		payload := map[string]interface{}{
			"embeds": []map[string]interface{}{
				{
					"title":       "\u274c Gralph Failed",
					"description": discordDescription,
					"color":       15548997,
					"fields": []map[string]interface{}{
						{
							"name":   "Project",
							"value":  fmt.Sprintf("`%s`", project),
							"inline": false,
						},
						{
							"name":   "Reason",
							"value":  reason,
							"inline": true,
						},
						{
							"name":   "Iterations",
							"value":  fmt.Sprintf("%s/%s", iterations, maxIterations),
							"inline": true,
						},
						{
							"name":   "Remaining Tasks",
							"value":  remaining,
							"inline": true,
						},
						{
							"name":   "Duration",
							"value":  duration,
							"inline": true,
						},
					},
					"footer": map[string]interface{}{
						"text": "Gralph CLI",
					},
					"timestamp": timestamp,
				},
			},
		}
		return json.Marshal(payload)
	case WebhookSlack:
		payload := map[string]interface{}{
			"attachments": []map[string]interface{}{
				{
					"color": "#ED4245",
					"blocks": []map[string]interface{}{
						{
							"type": "header",
							"text": map[string]interface{}{
								"type":  "plain_text",
								"text":  "\u274c Gralph Failed",
								"emoji": true,
							},
						},
						{
							"type": "section",
							"text": map[string]interface{}{
								"type": "mrkdwn",
								"text": slackDescription,
							},
						},
						{
							"type": "section",
							"fields": []map[string]interface{}{
								{
									"type": "mrkdwn",
									"text": fmt.Sprintf("*Project:*\n`%s`", project),
								},
								{
									"type": "mrkdwn",
									"text": fmt.Sprintf("*Reason:*\n%s", reason),
								},
								{
									"type": "mrkdwn",
									"text": fmt.Sprintf("*Iterations:*\n%s/%s", iterations, maxIterations),
								},
								{
									"type": "mrkdwn",
									"text": fmt.Sprintf("*Remaining Tasks:*\n%s", remaining),
								},
								{
									"type": "mrkdwn",
									"text": fmt.Sprintf("*Duration:*\n%s", duration),
								},
							},
						},
						{
							"type": "context",
							"elements": []map[string]interface{}{
								{
									"type": "mrkdwn",
									"text": fmt.Sprintf("Gralph CLI \u2022 %s", timestamp),
								},
							},
						},
					},
				},
			},
		}
		return json.Marshal(payload)
	default:
		payload := map[string]interface{}{
			"event":           "failed",
			"status":          "failure",
			"session":         opts.SessionName,
			"project":         project,
			"reason":          reason,
			"iterations":      iterations,
			"max_iterations":  maxIterations,
			"remaining_tasks": remaining,
			"duration":        duration,
			"timestamp":       timestamp,
			"message":         message,
		}
		return json.Marshal(payload)
	}
}

func formatDuration(duration time.Duration) string {
	if duration <= 0 {
		return "unknown"
	}
	total := int(duration.Seconds())
	if total <= 0 {
		return "unknown"
	}
	hours := total / 3600
	mins := (total % 3600) / 60
	secs := total % 60
	if hours > 0 {
		return fmt.Sprintf("%dh %dm %ds", hours, mins, secs)
	}
	if mins > 0 {
		return fmt.Sprintf("%dm %ds", mins, secs)
	}
	return fmt.Sprintf("%ds", secs)
}

func numberString(value int) string {
	if value <= 0 {
		return "unknown"
	}
	return strconv.Itoa(value)
}

func defaultString(value, fallback string) string {
	trimmed := strings.TrimSpace(value)
	if trimmed == "" {
		return fallback
	}
	return trimmed
}

func failedDescription(reason, sessionName string, slack bool) string {
	prefix := "**"
	suffix := "**"
	if slack {
		prefix = "*"
		suffix = "*"
	}
	label := fmt.Sprintf("%s%s%s", prefix, sessionName, suffix)
	switch reason {
	case "max_iterations":
		return fmt.Sprintf("Session %s hit maximum iterations limit.", label)
	case "error":
		return fmt.Sprintf("Session %s encountered an error.", label)
	case "manual_stop":
		return fmt.Sprintf("Session %s was manually stopped.", label)
	default:
		return fmt.Sprintf("Session %s failed: %s", label, reason)
	}
}

func failedMessage(reason, sessionName, iterations, maxIterations, remaining string) string {
	switch reason {
	case "max_iterations":
		return fmt.Sprintf("Gralph loop '%s' failed: hit max iterations (%s/%s) with %s tasks remaining", sessionName, iterations, maxIterations, remaining)
	case "error":
		return fmt.Sprintf("Gralph loop '%s' failed due to an error after %s iterations", sessionName, iterations)
	case "manual_stop":
		return fmt.Sprintf("Gralph loop '%s' was manually stopped after %s iterations with %s tasks remaining", sessionName, iterations, remaining)
	default:
		return fmt.Sprintf("Gralph loop '%s' failed: %s after %s iterations", sessionName, reason, iterations)
	}
}

File: gralph-rs/src/notify.rs
use reqwest::blocking::Client;
use serde_json::json;
use std::error::Error;
use std::fmt;
use std::time::Duration;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum WebhookType {
    Discord,
    Slack,
    Generic,
}

#[derive(Debug)]
pub enum NotifyError {
    InvalidInput(String),
    Http(reqwest::Error),
    HttpStatus(u16),
}

impl fmt::Display for NotifyError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            NotifyError::InvalidInput(message) => write!(f, "invalid input: {}", message),
            NotifyError::Http(err) => write!(f, "http error: {}", err),
            NotifyError::HttpStatus(code) => write!(f, "webhook returned HTTP {}", code),
        }
    }
}

impl Error for NotifyError {
    fn source(&self) -> Option<&(dyn Error + 'static)> {
        match self {
            NotifyError::Http(err) => Some(err),
            _ => None,
        }
    }
}

impl From<reqwest::Error> for NotifyError {
    fn from(error: reqwest::Error) -> Self {
        NotifyError::Http(error)
    }
}

pub fn detect_webhook_type(url: &str) -> WebhookType {
    let lower = url.to_lowercase();
    if lower.contains("discord.com/api/webhooks") || lower.contains("discordapp.com/api/webhooks") {
        WebhookType::Discord
    } else if lower.contains("hooks.slack.com") {
        WebhookType::Slack
    } else {
        WebhookType::Generic
    }
}

pub fn notify_complete(
    session_name: &str,
    webhook_url: &str,
    project_dir: Option<&str>,
    iterations: Option<u32>,
    duration_secs: Option<u64>,
    timeout_secs: Option<u64>,
) -> Result<(), NotifyError> {
    if session_name.trim().is_empty() {
        return Err(NotifyError::InvalidInput(
            "session name is required".to_string(),
        ));
    }
    if webhook_url.trim().is_empty() {
        return Err(NotifyError::InvalidInput(
            "webhook url is required".to_string(),
        ));
    }

    let project_dir = project_dir.unwrap_or("unknown");
    let iterations = iterations
        .map(|value| value.to_string())
        .unwrap_or_else(|| "unknown".to_string());
    let duration_str = format_duration(duration_secs);
    let timestamp = timestamp_iso8601();
    let webhook_type = detect_webhook_type(webhook_url);
    let payload = match webhook_type {
        WebhookType::Discord => format_discord_complete(
            session_name,
            project_dir,
            &iterations,
            &duration_str,
            &timestamp,
        ),
        WebhookType::Slack => format_slack_complete(
            session_name,
            project_dir,
            &iterations,
            &duration_str,
            &timestamp,
        ),
        WebhookType::Generic => format_generic_complete(
            session_name,
            project_dir,
            &iterations,
            &duration_str,
            &timestamp,
        ),
    }?;

    send_webhook(webhook_url, &payload, timeout_secs)
}

pub fn notify_failed(
    session_name: &str,
    webhook_url: &str,
    failure_reason: Option<&str>,
    project_dir: Option<&str>,
    iterations: Option<u32>,
    max_iterations: Option<u32>,
    remaining_tasks: Option<u32>,
    duration_secs: Option<u64>,
    timeout_secs: Option<u64>,
) -> Result<(), NotifyError> {
    if session_name.trim().is_empty() {
        return Err(NotifyError::InvalidInput(
            "session name is required".to_string(),
        ));
    }
    if webhook_url.trim().is_empty() {
        return Err(NotifyError::InvalidInput(
            "webhook url is required".to_string(),
        ));
    }

    let failure_reason = failure_reason.unwrap_or("unknown");
    let project_dir = project_dir.unwrap_or("unknown");
    let iterations = iterations
        .map(|value| value.to_string())
        .unwrap_or_else(|| "unknown".to_string());
    let max_iterations = max_iterations
        .map(|value| value.to_string())
        .unwrap_or_else(|| "unknown".to_string());
    let remaining_tasks = remaining_tasks
        .map(|value| value.to_string())
        .unwrap_or_else(|| "unknown".to_string());
    let duration_str = format_duration(duration_secs);
    let timestamp = timestamp_iso8601();

    let webhook_type = detect_webhook_type(webhook_url);
    let payload = match webhook_type {
        WebhookType::Discord => format_discord_failed(
            session_name,
            project_dir,
            failure_reason,
            &iterations,
            &max_iterations,
            &remaining_tasks,
            &duration_str,
            &timestamp,
        ),
        WebhookType::Slack => format_slack_failed(
            session_name,
            project_dir,
            failure_reason,
            &iterations,
            &max_iterations,
            &remaining_tasks,
            &duration_str,
            &timestamp,
        ),
        WebhookType::Generic => format_generic_failed(
            session_name,
            project_dir,
            failure_reason,
            &iterations,
            &max_iterations,
            &remaining_tasks,
            &duration_str,
            &timestamp,
        ),
    }?;

    send_webhook(webhook_url, &payload, timeout_secs)
}

pub fn send_webhook(
    url: &str,
    payload: &str,
    timeout_secs: Option<u64>,
) -> Result<(), NotifyError> {
    if url.trim().is_empty() {
        return Err(NotifyError::InvalidInput(
            "webhook url is required".to_string(),
        ));
    }
    if payload.trim().is_empty() {
        return Err(NotifyError::InvalidInput("payload is required".to_string()));
    }

    let timeout = timeout_secs.filter(|value| *value > 0).unwrap_or(30);
    let client = Client::builder()
        .timeout(Duration::from_secs(timeout))
        .build()?;
    let response = client
        .post(url)
        .header("Content-Type", "application/json")
        .body(payload.to_string())
        .send()?;

    if response.status().is_success() {
        Ok(())
    } else {
        Err(NotifyError::HttpStatus(response.status().as_u16()))
    }
}

fn format_discord_complete(
    session_name: &str,
    project_dir: &str,
    iterations: &str,
    duration_str: &str,
    timestamp: &str,
) -> Result<String, NotifyError> {
    let payload = json!({
        "embeds": [{
            "title": "✅ Gralph Complete",
            "description": format!("Session **{}** has finished all tasks successfully.", session_name),
            "color": 5763719,
            "fields": [
                {
                    "name": "Project",
                    "value": format!("`{}`", project_dir),
                    "inline": false
                },
                {
                    "name": "Iterations",
                    "value": iterations,
                    "inline": true
                },
                {
                    "name": "Duration",
                    "value": duration_str,
                    "inline": true
                }
            ],
            "footer": {
                "text": "Gralph CLI"
            },
            "timestamp": timestamp
        }]
    });
    Ok(serde_json::to_string_pretty(&payload)?)
}

fn format_slack_complete(
    session_name: &str,
    project_dir: &str,
    iterations: &str,
    duration_str: &str,
    timestamp: &str,
) -> Result<String, NotifyError> {
    let payload = json!({
        "attachments": [{
            "color": "#57F287",
            "blocks": [
                {
                    "type": "header",
                    "text": {
                        "type": "plain_text",
                        "text": "✅ Gralph Complete",
                        "emoji": true
                    }
                },
                {
                    "type": "section",
                    "text": {
                        "type": "mrkdwn",
                        "text": format!("Session *{}* has finished all tasks successfully.", session_name)
                    }
                },
                {
                    "type": "section",
                    "fields": [
                        {
                            "type": "mrkdwn",
                            "text": format!("*Project:*\n`{}`", project_dir)
                        },
                        {
                            "type": "mrkdwn",
                            "text": format!("*Iterations:*\n{}", iterations)
                        },
                        {
                            "type": "mrkdwn",
                            "text": format!("*Duration:*\n{}", duration_str)
                        }
                    ]
                },
                {
                    "type": "context",
                    "elements": [{
                        "type": "mrkdwn",
                        "text": format!("Gralph CLI • {}", timestamp)
                    }]
                }
            ]
        }]
    });
    Ok(serde_json::to_string_pretty(&payload)?)
}

fn format_slack_failed(
    session_name: &str,
    project_dir: &str,
    failure_reason: &str,
    iterations: &str,
    max_iterations: &str,
    remaining_tasks: &str,
    duration_str: &str,
    timestamp: &str,
) -> Result<String, NotifyError> {
    let description = match failure_reason {
        "max_iterations" => format!("Session *{}* hit maximum iterations limit.", session_name),
        "error" => format!("Session *{}* encountered an error.", session_name),
        "manual_stop" => format!("Session *{}* was manually stopped.", session_name),
        _ => format!("Session *{}* failed: {}", session_name, failure_reason),
    };

    let payload = json!({
        "attachments": [{
            "color": "#ED4245",
            "blocks": [
                {
                    "type": "header",
                    "text": {
                        "type": "plain_text",
                        "text": "❌ Gralph Failed",
                        "emoji": true
                    }
                },
                {
                    "type": "section",
                    "text": {
                        "type": "mrkdwn",
                        "text": description
                    }
                },
                {
                    "type": "section",
                    "fields": [
                        {
                            "type": "mrkdwn",
                            "text": format!("*Project:*\n`{}`", project_dir)
                        },
                        {
                            "type": "mrkdwn",
                            "text": format!("*Reason:*\n{}", failure_reason)
                        },
                        {
                            "type": "mrkdwn",
                            "text": format!("*Iterations:*\n{}/{}", iterations, max_iterations)
                        },
                        {
                            "type": "mrkdwn",
                            "text": format!("*Remaining Tasks:*\n{}", remaining_tasks)
                        },
                        {
                            "type": "mrkdwn",
                            "text": format!("*Duration:*\n{}", duration_str)
                        }
                    ]
                },
                {
                    "type": "context",
                    "elements": [{
                        "type": "mrkdwn",
                        "text": format!("Gralph CLI • {}", timestamp)
                    }]
                }
            ]
        }]
    });
    Ok(serde_json::to_string_pretty(&payload)?)
}

fn format_discord_failed(
    session_name: &str,
    project_dir: &str,
    failure_reason: &str,
    iterations: &str,
    max_iterations: &str,
    remaining_tasks: &str,
    duration_str: &str,
    timestamp: &str,
) -> Result<String, NotifyError> {
    let description = match failure_reason {
        "max_iterations" => format!("Session **{}** hit maximum iterations limit.", session_name),
        "error" => format!("Session **{}** encountered an error.", session_name),
        "manual_stop" => format!("Session **{}** was manually stopped.", session_name),
        _ => format!("Session **{}** failed: {}", session_name, failure_reason),
    };

    let payload = json!({
        "embeds": [{
            "title": "❌ Gralph Failed",
            "description": description,
            "color": 15548997,
            "fields": [
                {
                    "name": "Project",
                    "value": format!("`{}`", project_dir),
                    "inline": false
                },
                {
                    "name": "Reason",
                    "value": failure_reason,
                    "inline": true
                },
                {
                    "name": "Iterations",
                    "value": format!("{}/{}", iterations, max_iterations),
                    "inline": true
                },
                {
                    "name": "Remaining Tasks",
                    "value": remaining_tasks,
                    "inline": true
                },
                {
                    "name": "Duration",
                    "value": duration_str,
                    "inline": true
                }
            ],
            "footer": {
                "text": "Gralph CLI"
            },
            "timestamp": timestamp
        }]
    });
    Ok(serde_json::to_string_pretty(&payload)?)
}

fn format_generic_complete(
    session_name: &str,
    project_dir: &str,
    iterations: &str,
    duration_str: &str,
    timestamp: &str,
) -> Result<String, NotifyError> {
    let message = format!(
        "Gralph loop '{}' completed successfully after {} iterations ({})",
        session_name, iterations, duration_str
    );
    let payload = json!({
        "event": "complete",
        "status": "success",
        "session": session_name,
        "project": project_dir,
        "iterations": iterations,
        "duration": duration_str,
        "timestamp": timestamp,
        "message": message
    });
    Ok(serde_json::to_string_pretty(&payload)?)
}

fn format_generic_failed(
    session_name: &str,
    project_dir: &str,
    failure_reason: &str,
    iterations: &str,
    max_iterations: &str,
    remaining_tasks: &str,
    duration_str: &str,
    timestamp: &str,
) -> Result<String, NotifyError> {
    let message = match failure_reason {
        "max_iterations" => format!(
            "Gralph loop '{}' failed: hit max iterations ({}/{}) with {} tasks remaining",
            session_name, iterations, max_iterations, remaining_tasks
        ),
        "error" => format!(
            "Gralph loop '{}' failed due to an error after {} iterations",
            session_name, iterations
        ),
        "manual_stop" => format!(
            "Gralph loop '{}' was manually stopped after {} iterations with {} tasks remaining",
            session_name, iterations, remaining_tasks
        ),
        _ => format!(
            "Gralph loop '{}' failed: {} after {} iterations",
            session_name, failure_reason, iterations
        ),
    };
    let payload = json!({
        "event": "failed",
        "status": "failure",
        "session": session_name,
        "project": project_dir,
        "reason": failure_reason,
        "iterations": iterations,
        "max_iterations": max_iterations,
        "remaining_tasks": remaining_tasks,
        "duration": duration_str,
        "timestamp": timestamp,
        "message": message
    });
    Ok(serde_json::to_string_pretty(&payload)?)
}

fn format_duration(duration_secs: Option<u64>) -> String {
    let Some(total) = duration_secs else {
        return "unknown".to_string();
    };
    let hours = total / 3600;
    let mins = (total % 3600) / 60;
    let secs = total % 60;
    if hours > 0 {
        format!("{}h {}m {}s", hours, mins, secs)
    } else if mins > 0 {
        format!("{}m {}s", mins, secs)
    } else {
        format!("{}s", secs)
    }
}

fn timestamp_iso8601() -> String {
    chrono::Local::now().to_rfc3339()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn detect_webhook_type_matches() {
        assert_eq!(
            detect_webhook_type("https://discord.com/api/webhooks/123"),
            WebhookType::Discord
        );
        assert_eq!(
            detect_webhook_type("https://discordapp.com/api/webhooks/123"),
            WebhookType::Discord
        );
        assert_eq!(
            detect_webhook_type("https://hooks.slack.com/services/123"),
            WebhookType::Slack
        );
        assert_eq!(
            detect_webhook_type("https://example.com/webhook"),
            WebhookType::Generic
        );
    }
}
